In [1]:
import pandas as pd
import numpy as np
import re

In [2]:
pd.options.display.max_columns = 200
pd.options.display.max_rows = 170

# Overview of Data 

In [3]:
df = pd.read_csv("data/4_year/2019GraduationRates.csv", index_col="UnitID")

In [4]:
df[df.index.duplicated()]

Empty DataFrame
Columns: [Institution Name, City location of institution (HD2019), Grand total (S2019_IS  All instructional staff), Grand total men (S2019_IS  All instructional staff), Grand total women (S2019_IS  All instructional staff), American Indian or Alaska Native total (S2019_IS  All instructional staff), American Indian or Alaska Native men (S2019_IS  All instructional staff), American Indian or Alaska Native women (S2019_IS  All instructional staff), Asian total (S2019_IS  All instructional staff), Asian men (S2019_IS  All instructional staff), Asian women (S2019_IS  All instructional staff), Black or African American total (S2019_IS  All instructional staff), Black or African American men (S2019_IS  All instructional staff), Black or African American women (S2019_IS  All instructional staff), Hispanic or Latino total (S2019_IS  All instructional staff), Hispanic or Latino men (S2019_IS  All instructional staff), Hispanic or Latino women (S2019_IS  All instructional staff), Native Hawaiian or Other Pacific Islander total (S2019_IS  All instructional staff), Native Hawaiian or Other Pacific Islander men (S2019_IS  All instructional staff), Native Hawaiian or Other Pacific Islander women (S2019_IS  All instructional staff), White total (S2019_IS  All instructional staff), White men (S2019_IS  All instructional staff), White women (S2019_IS  All instructional staff), Two or more races total (S2019_IS  All instructional staff), Two or more races men (S2019_IS  All instructional staff), Two or more races women (S2019_IS  All instructional staff), Race/ethnicity unknown total (S2019_IS  All instructional staff), Race/ethnicity unknown men (S2019_IS  All instructional staff), Race/ethnicity unknown women (S2019_IS  All instructional staff), Nonresident alien total (S2019_IS  All instructional staff), Nonresident alien men (S2019_IS  All instructional staff), Nonresident alien women (S2019_IS  All instructional staff), Graduation rate  total cohort (DRVGR2019), Graduation rate  men (DRVGR2019), Graduation rate  women (DRVGR2019), Graduation rate  American Indian or Alaska Native (DRVGR2019), Graduation rate  Asian/Native Hawaiian/Other Pacific Islander (DRVGR2019), Graduation rate  Asian (DRVGR2019), Graduation rate  Native Hawaiian or Other Pacific Islander (DRVGR2019), Graduation rate  Black  non-Hispanic (DRVGR2019), Graduation rate  Hispanic (DRVGR2019), Graduation rate  White  non-Hispanic (DRVGR2019), Graduation rate  two or more races (DRVGR2019), Graduation rate  Race/ethnicity unknown (DRVGR2019), Graduation rate  Nonresident alien (DRVGR2019), Transfer-out rate  total cohort (DRVGR2019), Percent of full-time first-time undergraduates awarded any financial aid (SFA1819), Percent of full-time first-time undergraduates awarded federal  state  local or institutional grant aid (SFA1819), Core revenues  total dollars (GASB) (DRVF2019), Tuition and fees as a percent of core revenues (GASB) (DRVF2019), State appropriations as percent of core revenues  (GASB) (DRVF2019), Local appropriations as a percent of core revenues (GASB) (DRVF2019), Government grants and contracts as a percent of core revenues (GASB) (DRVF2019), Private gifts  grants  and contracts as a percent of core revenues (GASB) (DRVF2019), Investment return as a percent of core revenues (GASB) (DRVF2019), Other revenues as a percent of core revenues (GASB) (DRVF2019), Core revenues  total dollars (FASB) (DRVF2019), Tuition and fees as a percent of core revenues (FASB) (DRVF2019), Government grants and contracts as a percent of core revenues (FASB) (DRVF2019), Private gifts  grants  contracts/contributions from affiliated entities as a percent of core revenues (FASB) (DRVF2019), Investment return as a percent of core revenues (FASB) (DRVF2019), Other revenues as a percent of core revenues (FASB) (DRVF2019), Core revenues  total dollars (for-profit institutions) (DRVF2019), Tuition and fees as a percent of core revenues (for-profit institutions) (DRVF2019), Gov

In [5]:
df.drop(
    df.filter(regex="Unnamed").columns[0],
    axis=1, 
    inplace=True
)


In [6]:
df.describe()

Grand total (S2019_IS  All instructional staff)  \
count                                      2466.000000   
mean                                        259.922952   
std                                         533.437627   
min                                           1.000000   
25%                                          31.000000   
50%                                          91.000000   
75%                                         229.750000   
max                                        6787.000000   

       Grand total men (S2019_IS  All instructional staff)  \
count                                        2466.000000     
mean                                          139.502839     
std                                           303.402950     
min                                             0.000000     
25%                                            14.000000     
50%                                            44.000000     
75%                                           115.000000     
max                                          3983.000000     

       Grand total women (S2019_IS  All instructional staff)  \
count                                        2466.000000       
mean                                          120.420114       
std                                           235.195758       
min                                             0.000000       
25%                                            15.000000       
50%                                            46.000000       
75%                                           117.000000       
max                                          2804.000000       

       American Indian or Alaska Native total (S2019_IS  All instructional staff)  \
count                                        2466.000000                            
mean                                            0.969992                            
std                                             2.815905                            
min                                             0.000000                            
25%                                             0.000000                            
50%                                             0.000000                            
75%                                             1.000000                            
max                                            39.000000                            

       American Indian or Alaska Native men (S2019_IS  All instructional staff)  \
count                                        2466.000000                          
mean                                            0.467153                          
std                                             1.452959                          
min                                             0.000000                          
25%                                             0.000000                          
50%                                             0.000000                          
75%                                             0.000000                          
max                                            25.000000                          

       American Indian or Alaska Native women (S2019_IS  All instructional staff)  \
count                                        2466.000000                            
mean                                            0.502839                            
std                                             1.586259                            
min                                             0.000000                            
25%                                             0.000000                            
50%                                             0.000000                            
75%                                             0.000000                            
max                                            27.000000                            

       Asian total (S2019_IS  All instructional staff)  \
count     

## Graduation Rates

In [7]:
df.filter(regex="Grad").describe().loc["mean"].sort_values()

Graduation rate  Black  non-Hispanic (DRVGR2019)                             41.239763
Graduation rate  American Indian or Alaska Native (DRVGR2019)                41.953737
Graduation rate  Hispanic (DRVGR2019)                                        47.868434
Graduation rate  men (DRVGR2019)                                             47.922598
Graduation rate  two or more races (DRVGR2019)                               48.453488
Graduation rate  Native Hawaiian or Other Pacific Islander (DRVGR2019)       48.973489
Graduation rate  Race/ethnicity unknown (DRVGR2019)                          50.320922
Graduation rate  total cohort (DRVGR2019)                                    51.535205
Graduation rate  White  non-Hispanic (DRVGR2019)                             54.559891
Graduation rate  women (DRVGR2019)                                           54.874547
Graduation rate  Asian/Native Hawaiian/Other Pacific Islander (DRVGR2019)    57.539639
Graduation rate  Nonresident alien (DRVGR20

## Schema

In [8]:
# read in our schema
schema = pd.read_csv("data/4_year/schema.csv")

In [9]:
# Make a Pandas groupby that groups by variable name
groups = schema.groupby(by="VariableName")

In [10]:
#view all of our categorical groups
groups.head(2)

VariableName Value  \
0                          State abbreviation (HD2018)    AL   
1                          State abbreviation (HD2018)    AK   
58                      Sector of institution (HD2018)     1   
59                      Sector of institution (HD2018)     2   
66   Historically Black College or University (HD2018)     1   
67   Historically Black College or University (HD2018)     2   
68   Carnegie Classification 2018: Enrollment Profi...     1   
69   Carnegie Classification 2018: Enrollment Profi...     2   
76   Carnegie Classification 2018: Size and Setting...     1   
77   Carnegie Classification 2018: Size and Setting...     2   
95        Carnegie Classification 2018: Basic (HD2018)     1   
96        Carnegie Classification 2018: Basic (HD2018)     2   
128                 Institution size category (HD2018)     1   
129                 Institution size category (HD2018)     2   
135       Parent/child indicator - Finance (FLAGS2019)     1   
136       Parent/child indicator - Finance (FLAGS2019)     2   

                                            ValueLabel  
0                                              Alabama  
1                                               Alaska  
58                             Public, 4-year or above  
59             Private not-for-profit, 4-year or above  
66                                                 Yes  
67                                                  No  
68                  Exclusively undergraduate two-year  
69                 Exclusively undergraduate four-year  
76                                Two-year, very small  
77                                     Two-year, small  
95   Associate's Colleges: High Transfer-High Tradi...  
96   Associate's Colleges: High Transfer-Mixed Trad...  
128                                        Under 1,000  
129                                      1,000 - 4,999  
135  Parent record - includes data from branch camp...  
136    Child record - data reported with parent campus

**African American And Native American mean graduation rates are  38 and 37 respectively**

# Overview of Missing Data

## Admissions Test Scores

In [11]:
# How many schools failed to report their SAT Scores
df.filter(regex="SAT|ACT").isna().sum()

SAT Critical Reading 75th percentile score (ADM2014_RV)    1680
SAT Writing 75th percentile score (ADM2014_RV)             2128
SAT Math 75th percentile score (ADM2014_RV)                1667
ACT Composite 75th percentile score (ADM2014_RV)           1622
ACT English 75th percentile score (ADM2014_RV)             1767
ACT Math 75th percentile score (ADM2014_RV)                1765
ACT Writing 75th percentile score (ADM2014_RV)             2549
dtype: int64

## Finances

In [12]:
# How many schools failed to report their Finances
df.filter(regex="revenues|expenses").isna().sum().sample(5)

Public service expenses as a percent of total core expenses (GASB) (DRVF2019)      2080
Research expenses as a percent of total core expenses (GASB) (DRVF2019)            2080
Instruction expenses as a percent of total core expenses (GASB) (DRVF2019)         2080
Academic support expenses as a percent of total core expenses (GASB) (DRVF2019)    2080
Instruction expenses as a percent of total core expenses (FASB) (DRVF2019)         1156
dtype: int64

Notice that GASB, FASB and for-profit institutions are different groups

## Percentage of Enrollment by Race

In [13]:
df.filter(regex="Percent of total").isna().sum()

Percent of total enrollment that are White (DRVEF2014_RV)                                        170
Percent of total enrollment that are Black or African American  (DRVEF2014_RV)                   170
Percent of total enrollment that are Hispanic/Latino (DRVEF2014_RV)                              170
Percent of total enrollment that are Asian/Native Hawaiian/Pacific Islander (DRVEF2014_RV)       170
Percent of total enrollment that are American Indian or Alaska Native (DRVEF2014_RV)             170
Percent of total enrollment that are Race/ethnicity unknown (DRVEF2014_RV)                       170
Percent of total enrollment that are Nonresident Alien (DRVEF2014_RV)                            170
Percent of total enrollment that are women (DRVEF2014_RV)                                        170
Percent of total enrollment that are two or more races (DRVEF2014_RV)                            170
Percent of total enrollment that are Asian (DRVEF2014_RV)                                  

## Staff Diversity

In [14]:
# Staff Diveristy NA counts
df.filter(regex="instructional").isna().sum().head()

Grand total (S2019_IS  All instructional staff)                               380
Grand total men (S2019_IS  All instructional staff)                           380
Grand total women (S2019_IS  All instructional staff)                         380
American Indian or Alaska Native total (S2019_IS  All instructional staff)    380
American Indian or Alaska Native men (S2019_IS  All instructional staff)      380
dtype: int64

## Categorical

In [15]:
df.filter(regex="State abb|Carnegie|Historically|Sector|size category|(FLAGS2019)").isna().sum()

Carnegie Classification 2018: Size and Setting (HD2018)      55
Carnegie Classification 2018: Enrollment Profile (HD2018)    55
Carnegie Classification 2018: Basic (HD2018)                 55
State abbreviation (HD2018)                                  55
Institution size category (HD2018)                           55
Sector of institution (HD2018)                               55
Historically Black College or University (HD2018)            55
Parent/child indicator - Finance (FLAGS2019)                 21
dtype: int64

# Cleanning for Categorical Variables

## Non-Location Categories

In [16]:
# Selecting categorical columns
categorical_df = df.filter(regex="Carnegie|Historically|Sector|size category|(FLAGS2019)").copy(deep=True)

In [17]:
categorical_df.head()

Carnegie Classification 2018: Size and Setting (HD2018)  \
UnitID                                                            
177834                                               18.0         
180203                                                1.0         
222178                                               14.0         
138558                                               10.0         
488031                                                6.0         

        Carnegie Classification 2018: Enrollment Profile (HD2018)  \
UnitID                                                              
177834                                                7.0           
180203                                                1.0           
222178                                                4.0           
138558                                                2.0           
488031                                                6.0           

        Carnegie Classification 2018: Basic (HD2018)  \
UnitID                                                 
177834                                          25.0   
180203                                          33.0   
222178                                          18.0   
138558                                          23.0   
488031                                          31.0   

        Institution size category (HD2018)  Sector of institution (HD2018)  \
UnitID                                                                       
177834                                 2.0                             2.0   
180203                                 1.0                             4.0   
222178                                 3.0                             2.0   
138558                                 2.0                             1.0   
488031                                 1.0                             3.0   

        Historically Black College or University (HD2018)  \
UnitID                                                      
177834                                                2.0   
180203                                                2.0   
222178                                                2.0   
138558                                                2.0   
488031                                                2.0   

        Parent/child indicator - Finance (FLAGS2019)  
UnitID                                                
177834                                          -2.0  
180203                                          -2.0  
222178                                          -2.0  
138558                                          -2.0  
488031                                          -2.0

In [18]:
# Using Groupby object from schema to fill categorical variables with their string values
for i in categorical_df.columns:
    val_label = groups.get_group(i)[["Value", "ValueLabel"]].transpose()

    val_label.columns = val_label.loc["Value"]

    val_label.drop("Value", inplace=True)

    map_ = val_label.loc["ValueLabel"].to_dict()

    map_ ={int(i): v for i,v in map_.items()}
    map_[-100] = "isMissing"
    
    categorical_df[i] = categorical_df[i].fillna(-100).astype(int).replace(map_)

In [19]:
# View the changes
categorical_df.head()

Carnegie Classification 2018: Size and Setting (HD2018)  \
UnitID                                                           
177834                  Exclusively graduate/professional        
180203                               Two-year, very small        
222178              Four-year, medium, highly residential        
138558            Four-year, small, primarily residential        
488031    Four-year, very small, primarily nonresidential        

       Carnegie Classification 2018: Enrollment Profile (HD2018)  \
UnitID                                                             
177834                               Exclusively graduate          
180203                 Exclusively undergraduate two-year          
222178                                 High undergraduate          
138558                Exclusively undergraduate four-year          
488031                                  Majority graduate          

             Carnegie Classification 2018: Basic (HD2018)  \
UnitID                                                      
177834  Special Focus Four-Year: Medical Schools & Cen...   
180203                                    Tribal Colleges   
222178  Master's Colleges & Universities: Larger Programs   
138558  Baccalaureate/Associate's Colleges: Mixed Bacc...   
488031               Special Focus Four-Year: Law Schools   

       Institution size category (HD2018)  \
UnitID                                      
177834                      1,000 - 4,999   
180203                        Under 1,000   
222178                      5,000 - 9,999   
138558                      1,000 - 4,999   
488031                        Under 1,000   

                 Sector of institution (HD2018)  \
UnitID                                            
177834  Private not-for-profit, 4-year or above   
180203                           Public, 2-year   
222178  Private not-for-profit, 4-year or above   
138558                  Public, 4-year or above   
488031      Private for-profit, 4-year or above   

       Historically Black College or University (HD2018)  \
UnitID                                                     
177834                                                No   
180203                                                No   
222178                                                No   
138558                                                No   
488031                                                No   

       Parent/child indicator - Finance (FLAGS2019)  
UnitID                                               
177834                               Not applicable  
180203                               Not applicable  
222178                               Not applicable  
138558                               Not applicable  
488031                               Not applicable

In [20]:
# Clean the column names by getting rid of the (HD2018)
categorical_df.rename(columns=lambda x: re.sub(r'\(.*\)', '', x), inplace=True)
categorical_df.rename(columns=lambda x: x.replace(r'  ', ' '), inplace=True)
categorical_df.rename(columns=lambda x: x.strip(), inplace=True)

In [21]:
categorical_df.head()

Carnegie Classification 2018: Size and Setting  \
UnitID                                                    
177834                Exclusively graduate/professional   
180203                             Two-year, very small   
222178            Four-year, medium, highly residential   
138558          Four-year, small, primarily residential   
488031  Four-year, very small, primarily nonresidential   

       Carnegie Classification 2018: Enrollment Profile  \
UnitID                                                    
177834                             Exclusively graduate   
180203               Exclusively undergraduate two-year   
222178                               High undergraduate   
138558              Exclusively undergraduate four-year   
488031                                Majority graduate   

                      Carnegie Classification 2018: Basic  \
UnitID                                                      
177834  Special Focus Four-Year: Medical Schools & Cen...   
180203                                    Tribal Colleges   
222178  Master's Colleges & Universities: Larger Programs   
138558  Baccalaureate/Associate's Colleges: Mixed Bacc...   
488031               Special Focus Four-Year: Law Schools   

       Institution size category                    Sector of institution  \
UnitID                                                                      
177834             1,000 - 4,999  Private not-for-profit, 4-year or above   
180203               Under 1,000                           Public, 2-year   
222178             5,000 - 9,999  Private not-for-profit, 4-year or above   
138558             1,000 - 4,999                  Public, 4-year or above   
488031               Under 1,000      Private for-profit, 4-year or above   

       Historically Black College or University  \
UnitID                                            
177834                                       No   
180203                                       No   
222178                                       No   
138558                                       No   
488031                                       No   

       Parent/child indicator - Finance  
UnitID                                   
177834                   Not applicable  
180203                   Not applicable  
222178                   Not applicable  
138558                   Not applicable  
488031                   Not applicable

## Location Data

In [22]:
# Selecting location data
location_data = df.filter(regex="State abbr|City").copy(deep=True)
location_data.fillna("isMissing", inplace=True)

# Financial Data

Universities report their finances to the Ipeds data center as either a GASB, FASB, or for-profit institution. Therefore FASB and for-profit institutions will have all null values in the GASB finance columns. I will clean this by preserving the type of school they are (FASB, GASB, or for-profit) whilst merging finance columns into one shared column for FASB, GASB and for-profit colleges.



In [23]:
gasb_df = df.filter(regex="(GASB)").copy(deep=True)

In [24]:
gasb_df.head()

Core revenues  total dollars (GASB) (DRVF2019)  \
UnitID                                                   
177834                                             NaN   
180203                                       7078959.0   
222178                                             NaN   
138558                                      70491879.0   
488031                                             NaN   

        Tuition and fees as a percent of core revenues (GASB) (DRVF2019)  \
UnitID                                                                     
177834                                                NaN                  
180203                                                2.0                  
222178                                                NaN                  
138558                                               17.0                  
488031                                                NaN                  

        State appropriations as percent of core revenues  (GASB) (DRVF2019)  \
UnitID                                                                        
177834                                                NaN                     
180203                                                1.0                     
222178                                                NaN                     
138558                                               37.0                     
488031                                                NaN                     

        Local appropriations as a percent of core revenues (GASB) (DRVF2019)  \
UnitID                                                                         
177834                                                NaN                      
180203                                                0.0                      
222178                                                NaN                      
138558                                                0.0                      
488031                                                NaN                      

        Government grants and contracts as a percent of core revenues (GASB) (DRVF2019)  \
UnitID                                                                                    
177834                                                NaN                                 
180203                                               60.0                                 
222178                                                NaN                                 
138558                                               17.0                                 
488031                                                NaN                                 

        Private gifts  grants  and contracts as a percent of core revenues (GASB) (DRVF2019)  \
UnitID                                                                                         
177834                                                NaN                                      
180203                                                7.0                                      
222178                                                NaN                                      
138558                                                2.0                                      
488031                                                NaN                                      

        Investment return as a percent of core revenues (GASB) (DRVF2019)  \
UnitID                                                                      
177834                                                NaN                   
180203                                                1.0                   
222178                                                NaN                   
138558                                                0.0                   
488031                                                NaN                   

        Other revenues as a percent of core revenues (GASB) (DRVF2019)  \
UnitID                      

In [25]:
fasb_df = df.filter(regex="(FASB)").copy(deep=True)

In [26]:
fasb_df.head()

Core revenues  total dollars (FASB) (DRVF2019)  \
UnitID                                                   
177834                                     174782780.0   
180203                                             NaN   
222178                                     136379482.0   
138558                                             NaN   
488031                                             NaN   

        Tuition and fees as a percent of core revenues (FASB) (DRVF2019)  \
UnitID                                                                     
177834                                               80.0                  
180203                                                NaN                  
222178                                               57.0                  
138558                                                NaN                  
488031                                                NaN                  

        Government grants and contracts as a percent of core revenues (FASB) (DRVF2019)  \
UnitID                                                                                    
177834                                                2.0                                 
180203                                                NaN                                 
222178                                                4.0                                 
138558                                                NaN                                 
488031                                                NaN                                 

        Private gifts  grants  contracts/contributions from affiliated entities as a percent of core revenues (FASB) (DRVF2019)  \
UnitID                                                                                                                            
177834                                                6.0                                                                         
180203                                                NaN                                                                         
222178                                               17.0                                                                         
138558                                                NaN                                                                         
488031                                                NaN                                                                         

        Investment return as a percent of core revenues (FASB) (DRVF2019)  \
UnitID                                                                      
177834                                                6.0                   
180203                                                NaN                   
222178                                               18.0                   
138558                                                NaN                   
488031                                                NaN                   

        Other revenues as a percent of core revenues (FASB) (DRVF2019)  \
UnitID                                                                   
177834                                                6.0                
180203                                                NaN                
222178                                                4.0                
138558                                                NaN                
488031                                                NaN                

        Core expenses  total dollars (FASB) (DRVF2019)  \
UnitID                                                   
177834                                     148677539.0   
180203                                             NaN   
222178                                     121513039.0   
138558                                             NaN   
488031                                             NaN   

        Instruction expenses as a percent of t

In [27]:
for_profit = df.filter(regex="(for-profit institut)").copy(deep=True)

In [28]:
for_profit.head()

Core revenues  total dollars (for-profit institutions) (DRVF2019)  \
UnitID                                                                      
177834                                                NaN                   
180203                                                NaN                   
222178                                                NaN                   
138558                                                NaN                   
488031                                          1555377.0                   

        Tuition and fees as a percent of core revenues (for-profit institutions) (DRVF2019)  \
UnitID                                                                                        
177834                                                NaN                                     
180203                                                NaN                                     
222178                                                NaN                                     
138558                                                NaN                                     
488031                                               99.0                                     

        Government grants and contracts as a percent of core revenues (for-profit institutions) (DRVF2019)  \
UnitID                                                                                                       
177834                                                NaN                                                    
180203                                                NaN                                                    
222178                                                NaN                                                    
138558                                                NaN                                                    
488031                                                0.0                                                    

        Private gifts  grants  contracts as a percent of core revenues (for-profit institutions) (DRVF2019)  \
UnitID                                                                                                        
177834                                                NaN                                                     
180203                                                NaN                                                     
222178                                                NaN                                                     
138558                                                NaN                                                     
488031                                                0.0                                                     

        Investment return as a percent of core revenues (for-profit institutions) (DRVF2019)  \
UnitID                                                                                         
177834                                                NaN                                      
180203                                                NaN                                      
222178                                                NaN                                      
138558                                                NaN                                      
488031                                                1.0                                      

        Sales and services of educational activities as a percent of core revenues (for-profit institutions) (DRVF2019)  \
UnitID                                                                                                                    
177834                                                NaN                                                                 
180203                                                NaN                                                                 
222178                                                NaN                               

In [29]:
# create GASB, FASB and forProfit columns
gasb_df["is_GASB"] = gasb_df.sum(axis=1).astype(int).apply(lambda x: 1 if x != 0 else 0)
gasb_df["is_FASB"] = 0
gasb_df["is_forProfit"] = 0

fasb_df["is_GASB"] = 0
fasb_df["is_FASB"] = fasb_df.sum(axis=1).astype(int).apply(lambda x: 1 if x != 0 else 0)
fasb_df["is_forProfit"] = 0

for_profit["is_GASB"] = 0
for_profit["is_FASB"] = 0
for_profit["is_forProfit"] = for_profit.sum(axis=1).astype(int).apply(lambda x: 1 if x != 0 else 0)

In [30]:
# View how many of each typ of schoo their are
len(for_profit[for_profit.is_forProfit == 1])

358

In [31]:
gasb_df.columns[2:4]

Index(['State appropriations as percent of core revenues  (GASB) (DRVF2019)', 'Local appropriations as a percent of core revenues (GASB) (DRVF2019)'], dtype='object')

In [32]:
# Separating out State and Local Appropriations which are unique to gasb
state_local = gasb_df[gasb_df.columns[2:4]]
gasb_df = gasb_df.drop(gasb_df.columns[2:4], axis=1).copy(deep=True)

In [33]:
state_local.rename(columns=lambda x: re.sub(r'\(.*\)', '', x), inplace=True)
state_local.rename(columns=lambda x: x.replace(r'  ', ' '), inplace=True)
state_local.rename(columns=lambda x: x.strip(), inplace=True)

## Merging GASB, FASB and for-profit institutions Finances into one column

In order to merge GASB, FASB, and for-profit institutions we need for all of their column names to be the same value

In [34]:

fasb_df.rename(columns=lambda x: re.sub(r'\(.*\)', '', x), inplace=True)
fasb_df.rename(columns=lambda x: x.replace(r'  ', ' '), inplace=True)
fasb_df.rename(columns=lambda x: x.strip(), inplace=True)

fasb_df = fasb_df[sorted(fasb_df.columns)]

In [35]:
gasb_df.rename(columns=lambda x: re.sub(r'\(.*\)', '', x), inplace=True)
gasb_df.rename(columns=lambda x: x.replace(r'  ', ' '), inplace=True)
gasb_df.rename(columns=lambda x: x.strip(), inplace=True)

gasb_df = gasb_df[sorted(gasb_df.columns)]

In [36]:
for_profit.rename(columns=lambda x: re.sub(r'\(.*\)', '', x), inplace=True)
for_profit.rename(columns=lambda x: x.replace(r'  ', ' '), inplace=True)
for_profit.rename(columns=lambda x: x.strip(), inplace=True)

# Typo in column name government grants and contracts
for_profit.rename(columns=lambda x: re.sub(r'grants contracts', 'grants and contracts', x), inplace=True)
for_profit = for_profit[sorted(for_profit.columns)]

Lets view the data frames

In [37]:
for_profit.head(2)

Academic support expenses as a percent of total core expenses  \
UnitID                                                                  
177834                                                NaN               
180203                                                NaN               

        Core expenses total dollars  Core revenues total dollars  \
UnitID                                                             
177834                          NaN                          NaN   
180203                          NaN                          NaN   

        Government grants and contracts as a percent of core revenues  \
UnitID                                                                  
177834                                                NaN               
180203                                                NaN               

        Institutional support expenses as a percent of total core expenses  \
UnitID                                                                       
177834                                                NaN                    
180203                                                NaN                    

        Instruction expenses as a percent of total core expenses  \
UnitID                                                             
177834                                                NaN          
180203                                                NaN          

        Investment return as a percent of core revenues  \
UnitID                                                    
177834                                              NaN   
180203                                              NaN   

        Other core expenses as a percent of total core expenses  \
UnitID                                                            
177834                                                NaN         
180203                                                NaN         

        Other revenues as a percent of core revenues  \
UnitID                                                 
177834                                           NaN   
180203                                           NaN   

        Private gifts grants and contracts as a percent of core revenues  \
UnitID                                                                     
177834                                                NaN                  
180203                                                NaN                  

        Public service expenses as a percent of total core expenses  \
UnitID                                                                
177834                                                NaN             
180203                                                NaN             

        Research expenses as a percent of total core expenses  \
UnitID                                                          
177834                                                NaN       
180203                                                NaN       

        Sales and services of educational activities as a percent of core revenues  \
UnitID                                                                               
177834                                                NaN                            
180203                                                NaN                            

        Student service expenses as a percent of total core expenses  \
UnitID                                                                 
177834                                                NaN              
180203                                                NaN              

        Tuition and fees as a percent of core revenues  is_FASB  is_GASB  \
UnitID                                                                     
177834                                             NaN        0        0   
180203                                             NaN        0        0   

        is_forProfit  
UnitID             

In [38]:
gasb_df.head(2)

Academic support expenses as a percent of total core expenses  \
UnitID                                                                  
177834                                                NaN               
180203                                                3.0               

        Core expenses total dollars  Core revenues total dollars  \
UnitID                                                             
177834                          NaN                          NaN   
180203                    7294344.0                    7078959.0   

        Government grants and contracts as a percent of core revenues  \
UnitID                                                                  
177834                                                NaN               
180203                                               60.0               

        Institutional support expenses as a percent of total core expenses  \
UnitID                                                                       
177834                                                NaN                    
180203                                               16.0                    

        Instruction expenses as a percent of total core expenses  \
UnitID                                                             
177834                                                NaN          
180203                                               29.0          

        Investment return as a percent of core revenues  \
UnitID                                                    
177834                                              NaN   
180203                                              1.0   

        Other core expenses as a percent of total core expenses  \
UnitID                                                            
177834                                                NaN         
180203                                               36.0         

        Other revenues as a percent of core revenues  \
UnitID                                                 
177834                                           NaN   
180203                                          29.0   

        Private gifts grants and contracts as a percent of core revenues  \
UnitID                                                                     
177834                                                NaN                  
180203                                                7.0                  

        Public service expenses as a percent of total core expenses  \
UnitID                                                                
177834                                                NaN             
180203                                                7.0             

        Research expenses as a percent of total core expenses  \
UnitID                                                          
177834                                                NaN       
180203                                                3.0       

        Student service expenses as a percent of total core expenses  \
UnitID                                                                 
177834                                                NaN              
180203                                                8.0              

        Tuition and fees as a percent of core revenues  is_FASB  is_GASB  \
UnitID                                                                     
177834                                             NaN        0        0   
180203                                             2.0        0        1   

        is_forProfit  
UnitID                
177834             0  
180203             0

In [39]:
fasb_df.head(2)

Academic support expenses as a percent of total core expenses  \
UnitID                                                                  
177834                                               27.0               
180203                                                NaN               

        Core expenses total dollars  Core revenues total dollars  \
UnitID                                                             
177834                  148677539.0                  174782780.0   
180203                          NaN                          NaN   

        Government grants and contracts as a percent of core revenues  \
UnitID                                                                  
177834                                                2.0               
180203                                                NaN               

        Institutional support expenses as a percent of total core expenses  \
UnitID                                                                       
177834                                               15.0                    
180203                                                NaN                    

        Instruction expenses as a percent of total core expenses  \
UnitID                                                             
177834                                               50.0          
180203                                                NaN          

        Investment return as a percent of core revenues  \
UnitID                                                    
177834                                              6.0   
180203                                              NaN   

        Other core expenses as a percent of total core expenses  \
UnitID                                                            
177834                                                1.0         
180203                                                NaN         

        Other revenues as a percent of core revenues  \
UnitID                                                 
177834                                           6.0   
180203                                           NaN   

        Private gifts grants contracts/contributions from affiliated entities as a percent of core revenues  \
UnitID                                                                                                        
177834                                                6.0                                                     
180203                                                NaN                                                     

        Public service expenses as a percent of total core expenses  \
UnitID                                                                
177834                                                0.0             
180203                                                NaN             

        Research expenses as a percent of total core expenses  \
UnitID                                                          
177834                                                3.0       
180203                                                NaN       

        Student service expenses as a percent of total core expenses  \
UnitID                                                                 
177834                                                4.0              
180203                                                NaN              

        Tuition and fees as a percent of core revenues  is_FASB  is_GASB  \
UnitID                                                                     
177834                                            80.0        1        0   
180203                                             NaN        0        0   

        is_forProfit  
UnitID                
177834             0  
180203             0

**Lets make sure each subset contains the same amount of Columns**

In [40]:
set(for_profit.columns) - set(gasb_df.columns)

{'Sales and services of educational activities as a percent of core revenues'}

In [41]:
# Viewing sales and services
print(for_profit.columns[-5])
print(for_profit.columns[-6])
print(for_profit[for_profit.columns[-6]].isna().sum())
print(for_profit[for_profit.columns[-5]].isna().sum())

Student service expenses as a percent of total core expenses
Sales and services of educational activities as a percent of core revenues
2488
2488


In [42]:
(for_profit[for_profit.columns[-5]] + for_profit[for_profit.columns[-6]]).isna().sum()

2488

In [43]:
# Student services = student services + sales and services
for_profit[for_profit.columns[-5]] = for_profit[for_profit.columns[-5]] + for_profit[for_profit.columns[-6]]

In [44]:
# drop sales and services
for_profit.drop(for_profit.columns[-6], axis=1, inplace=True)

In [45]:
drop_sales = 'Sales and services of educational activities as a percent of core revenues'

Renaming private grants and contracts so that is is the same for fasb, gasb and nasb

In [46]:
gasb_df.columns[9] # this column is different for gasb but they contain the same data

'Private gifts grants and contracts as a percent of core revenues'

In [47]:
fasb_df.columns[9] # I need to rename this column so that it is the same as GASB and for-profit institutions

'Private gifts grants contracts/contributions from affiliated entities as a percent of core revenues'

In [48]:
fasb_df.rename({fasb_df.columns[9]: gasb_df.columns[9]}, axis=1, inplace=True)

**Sales and services are only recorded for for-profit institutions so I will combine that column with other services so that Services include Sales and Services**

In [49]:
#check to see if all columns are equal
print("gasb == fasb: ", gasb_df.columns == fasb_df.columns)
print("fasb == for_profit: ", fasb_df.columns == for_profit.columns)

gasb == fasb:  [ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True]
fasb == for_profit:  [ True  True  True  True  True  True  True  True  True  True  True  True
  True  True  True  True  True]


In [50]:
#check 

fasb_df.iloc[:, -3:]

is_FASB  is_GASB  is_forProfit
UnitID                                
177834        1        0             0
180203        0        0             0
222178        1        0             0
138558        0        0             0
488031        0        0             0
...         ...      ...           ...
217059        1        0             0
141361        1        0             0
206695        0        0             0
204255        0        0             0
458575        0        0             0

[2846 rows x 3 columns]

### Finally Merging all together

In [51]:
finances_df = pd.concat([gasb_df, fasb_df, for_profit])

In [52]:
#check
finances_df.iloc[:, -3:]

is_FASB  is_GASB  is_forProfit
UnitID                                
177834        0        0             0
180203        0        1             0
222178        0        0             0
138558        0        1             0
488031        0        0             0
...         ...      ...           ...
217059        0        0             0
141361        0        0             0
206695        0        0             0
204255        0        0             0
458575        0        0             0

[8538 rows x 3 columns]

In [53]:
finances_df["isnan"] = finances_df.isna().sum(axis=1)

In [54]:
#check
finances_df.iloc[:, -4:]

is_FASB  is_GASB  is_forProfit  isnan
UnitID                                       
177834        0        0             0     14
180203        0        1             0      0
222178        0        0             0     14
138558        0        1             0      0
488031        0        0             0     14
...         ...      ...           ...    ...
217059        0        0             0     14
141361        0        0             0     14
206695        0        0             0     14
204255        0        0             0     14
458575        0        0             0     14

[8538 rows x 4 columns]

In [55]:
finances_df = finances_df[finances_df["isnan"] != 14].copy(deep=True)

In [56]:
#check
finances_df

Academic support expenses as a percent of total core expenses  \
UnitID                                                                  
180203                                                3.0               
138558                                               15.0               
126182                                                7.0               
100654                                                6.0               
100724                                                9.0               
...                                                   ...               
458229                                                5.0               
458210                                                5.0               
490133                                                8.0               
224660                                               23.0               
224679                                               25.0               

        Core expenses total dollars  Core revenues total dollars  \
UnitID                                                             
180203                    7294344.0                    7078959.0   
138558                   50201124.0                   70491879.0   
126182                   35725137.0                   52099218.0   
100654                  146878227.0                  143112797.0   
100724                  134888910.0                  134993142.0   
...                             ...                          ...   
458229                   40658057.0                   64468935.0   
458210                   61800247.0                   97992780.0   
490133                    7497616.0                    8972118.0   
224660                    9920436.0                    9912034.0   
224679                   14831336.0                   14868051.0   

        Government grants and contracts as a percent of core revenues  \
UnitID                                                                  
180203                                               60.0               
138558                                               17.0               
126182                                               51.0               
100654                                               30.0               
100724                                               25.0               
...                                                   ...               
458229                                                0.0               
458210                                                0.0               
490133                                                0.0               
224660                                               27.0               
224679                                               35.0               

        Institutional support expenses as a percent of total core expenses  \
UnitID                                                                       
180203                                               16.0                    
138558                                               19.0                    
126182                                               15.0                    
100654                                               11.0                    
100724                                               27.0                    
...                                                   ...                    
458229                                               43.0                    
458210                                               43.0                    
490133                                               63.0                    
224660                                               10.0                    
224679                                               13.0                    

        Instruction expenses as a percent of total core expenses  \
UnitID                                                             
180203                                               29.0          
1385

In [57]:
finances_df.drop("isnan", axis=1, inplace=True)

In [58]:
#changelog
finances_df = finances_df.reset_index().drop_duplicates(subset="UnitID")

In [59]:
finances_df.set_index("UnitID", inplace=True)

In [60]:
print(len(finances_df))
print(finances_df.isna().sum())

2814
Academic support expenses as a percent of total core expenses         0
Core expenses total dollars                                           0
Core revenues total dollars                                           0
Government grants and contracts as a percent of core revenues         1
Institutional support expenses as a percent of total core expenses    0
Instruction expenses as a percent of total core expenses              0
Investment return as a percent of core revenues                       1
Other core expenses as a percent of total core expenses               0
Other revenues as a percent of core revenues                          1
Private gifts grants and contracts as a percent of core revenues      1
Public service expenses as a percent of total core expenses           0
Research expenses as a percent of total core expenses                 0
Student service expenses as a percent of total core expenses          0
Tuition and fees as a percent of core revenues             

# Find any possible accidental Values

## Negative Values

In [61]:
neg_cols = finances_df.columns[finances_df[finances_df < 0].isna().sum() != 6197]

In [62]:
neg_cols

Index(['Academic support expenses as a percent of total core expenses',
       'Core expenses total dollars', 'Core revenues total dollars',
       'Government grants and contracts as a percent of core revenues',
       'Institutional support expenses as a percent of total core expenses',
       'Instruction expenses as a percent of total core expenses',
       'Investment return as a percent of core revenues',
       'Other core expenses as a percent of total core expenses',
       'Other revenues as a percent of core revenues',
       'Private gifts grants and contracts as a percent of core revenues',
       'Public service expenses as a percent of total core expenses',
       'Research expenses as a percent of total core expenses',
       'Student service expenses as a percent of total core expenses',
       'Tuition and fees as a percent of core revenues', 'is_FASB', 'is_GASB',
       'is_forProfit'],
      dtype='object')

In [63]:
neg_reporters = set()
for i in neg_cols:
    for j in finances_df[finances_df[i]< 0][i].index:
        neg_reporters.add(j)
    

In [64]:
len(neg_reporters)

109

## Lets have a look at these negative percentages

**These schools must have had some negative return on investment, so this information remains accurate and won't be tampered with** Lets Insure that their revenues add up to roughly %100 and their expenses do the same

In [65]:
finances_df["revenues_percentage_sum"] = finances_df.filter(regex="as a percent.*revenues").sum(axis=1)

In [66]:
finances_df["expenses_percentage_sum"] = finances_df.filter(regex="as a percent.*expenses").sum(axis=1)

In [67]:
finances_df[finances_df["revenues_percentage_sum"] < 75]

Academic support expenses as a percent of total core expenses  \
UnitID                                                                  
138558                                               15.0               
100654                                                6.0               
100724                                                9.0               
138716                                               10.0               
175342                                                9.0               
...                                                   ...               
446002                                                6.0               
429094                                               27.0               
475635                                                8.0               
480912                                               31.0               
493673                                                7.0               

        Core expenses total dollars  Core revenues total dollars  \
UnitID                                                             
138558                   50201124.0                   70491879.0   
100654                  146878227.0                  143112797.0   
100724                  134888910.0                  134993142.0   
138716                   89924505.0                   81754461.0   
175342                   76086915.0                   75226337.0   
...                             ...                          ...   
446002                    1970923.0                    2113565.0   
429094                    3490318.0                    3460747.0   
475635                     478928.0                     485447.0   
480912                     284370.0                     285527.0   
493673                     457305.0                     455785.0   

        Government grants and contracts as a percent of core revenues  \
UnitID                                                                  
138558                                               17.0               
100654                                               30.0               
100724                                               25.0               
138716                                               36.0               
175342                                               43.0               
...                                                   ...               
446002                                                0.0               
429094                                                1.0               
475635                                                0.0               
480912                                                0.0               
493673                                                0.0               

        Institutional support expenses as a percent of total core expenses  \
UnitID                                                                       
138558                                               19.0                    
100654                                               11.0                    
100724                                               27.0                    
138716                                               15.0                    
175342                                               20.0                    
...                                                   ...                    
446002                                                5.0                    
429094                                               22.0                    
475635                                               68.0                    
480912                                               19.0                    
493673                                               43.0                    

        Instruction expenses as a percent of total core expenses  \
UnitID                                                             
138558                                               42.0          
1006

### I am chosing to ignore the fact that some schools report revenue percentages that don't add up to 100%. I am guessing that there is a good reason for this i.e negative return on investment and that the information is still accurate and will help our model

In [68]:
finances_df.drop(finances_df.columns[-2:], axis=1, inplace=True)

## Final Merge

### State and Local Appropriations

In [69]:
 finances_df = state_local\
    .reset_index().merge(finances_df.reset_index(), on="UnitID", how="outer").set_index("UnitID")

In [70]:
finances_df.isna().sum()

State appropriations as percent of core revenues                      2080
Local appropriations as a percent of core revenues                    2080
Academic support expenses as a percent of total core expenses           32
Core expenses total dollars                                             32
Core revenues total dollars                                             32
Government grants and contracts as a percent of core revenues           33
Institutional support expenses as a percent of total core expenses      32
Instruction expenses as a percent of total core expenses                32
Investment return as a percent of core revenues                         33
Other core expenses as a percent of total core expenses                 32
Other revenues as a percent of core revenues                            33
Private gifts grants and contracts as a percent of core revenues        33
Public service expenses as a percent of total core expenses             32
Research expenses as a pe

### Returning Finances Back to the original df

# Final Merge

In [71]:
# Lets clean the column names the same way for all columns with regex
df.rename(columns=lambda x: re.sub(r'\(.*25.*over.*\)', 'Age 25 and over', x), inplace=True)
df.rename(columns=lambda x: re.sub(r'\(.*under.*\)', 'Age 25 and under', x), inplace=True)

df.rename(columns=lambda x: re.sub(r'\(.*instructional staff.*\)', 'instructional staff', x), inplace=True)
df.rename(columns=lambda x: re.sub(r'\(.*\)', '', x), inplace=True)
df.rename(columns=lambda x: x.replace(r'  ', ' '), inplace=True)
df.rename(columns=lambda x: x.strip(), inplace=True)

In [72]:
df = df.loc[:, ~df.columns.duplicated()]

In [73]:
financial_cols = df.filter(regex='revenues|expenses').columns

In [74]:
df.drop(financial_cols, axis=1, inplace=True)

In [75]:
# sort in alphabetical order by university name
df.sort_index(inplace=True)
finances_df.sort_index(inplace=True)

In [76]:
# Both data frames have the same exact indices
(finances_df.index == df.index).sum()

2846

In [77]:
# Its safe to concatenate side by side because we have verified that they have the same exact indices 
final_df = pd.concat([df, finances_df], axis=1)

In [78]:
# lets drop where total graduation rates are missing

In [79]:
grad_rates = final_df.filter(regex="Graduation").columns[0]

In [80]:
final_df.dropna(subset=[grad_rates], inplace=True)

In [81]:
final_df

Institution Name  \
UnitID                                                      
100654                           Alabama A & M University   
100663                University of Alabama at Birmingham   
100690                                 Amridge University   
100706                University of Alabama in Huntsville   
100724                           Alabama State University   
...                                                   ...   
492801  Drury University-College of Continuing Profess...   
492962  Indiana Institute of Technology-College of Pro...   
494597   Faith Theological Seminary and Christian College   
494603          Southwestern Baptist Theological Seminary   
494685                                     Urshan College   

       City location of institution  Grand total instructional staff  \
UnitID                                                                 
100654                       Normal                            242.0   
100663                   Birmingham                           2523.0   
100690                   Montgomery                             30.0   
100706                   Huntsville                            363.0   
100724                   Montgomery                            226.0   
...                             ...                              ...   
492801                  Springfield                              8.0   
492962                   Fort Wayne                              NaN   
494597                        Tampa                              NaN   
494603                   Fort Worth                              NaN   
494685                   Wentzville                              NaN   

        Grand total men instructional staff  \
UnitID                                        
100654                                131.0   
100663                               1441.0   
100690                                 24.0   
100706                                196.0   
100724                                114.0   
...                                     ...   
492801                                  2.0   
492962                                  NaN   
494597                                  NaN   
494603                                  NaN   
494685                                  NaN   

        Grand total women instructional staff  \
UnitID                                          
100654                                  111.0   
100663                                 1082.0   
100690                                    6.0   
100706                                  167.0   
100724                                  112.0   
...                                       ...   
492801                                    6.0   
492962                                    NaN   
494597                                    NaN   
494603                                    NaN   
494685                                    NaN   

        American Indian or Alaska Native total instructional staff  \
UnitID                                                               
100654                                                1.0            
100663                                                4.0            
100690                                                0.0            
100706                                                1.0            
100724                                                0.0            
...                                                   ...            
492801                                                0.0            
492962                                                NaN            
494597                                                NaN            
494603                                                NaN            
494685                                                NaN            

        American Indian or Alaska Native men instructional staff  \
UnitID                                                             
100654       

# Imputation Pipeline

In [82]:
#SAT scores have the highest missing value count
final_df.isna().sum().sort_values(ascending=False)

ACT Writing 75th percentile score                                                                        2020
SAT Writing 75th percentile score                                                                        1600
State appropriations as percent of core revenues                                                         1598
Local appropriations as a percent of core revenues                                                       1598
Graduation rate Native Hawaiian or Other Pacific Islander                                                1372
ACT English 75th percentile score                                                                        1240
ACT Math 75th percentile score                                                                           1238
SAT Critical Reading 75th percentile score                                                               1153
SAT Math 75th percentile score                                                                           1140
ACT Compos

In [83]:
test_scores = final_df.filter(regex='ACT|SAT').columns

We only need the school to have reported 1 of these scores to get a good idea about what the rest of the scores would be. First we will create isna columns for these values. Since  ACT Composite has the least amount of null values, we will use this to fill na values in other columns.

In [84]:
for i in test_scores:
    name = i + "_isNaN"
    final_df[name] = final_df[i].isna() * 1

ACT Scores range from 1-36 while SAT scores range from 1-800 so I will create a mapping from ACT to SAT scores using np.linspace

In [85]:
final_df.filter(regex='ACT').iloc[:, 0].unique()

array([19., 27., nan, 30., 31., 24., 29., 26., 23., 25., 22., 18., 28.,
       21., 35., 20., 34., 33., 32., 17., 16.])

In [86]:
A = np.linspace(1, 800, 37)
B = np.linspace(160, 0, 37)
C = A + B

In [87]:
act_to_sat = dict(enumerate(C))

In [88]:
# This mapping is close (not exac) to what I see online with score conversion approximations
print(f'a {28} on the ACT maps to a {act_to_sat[28]} on the SAT')
print(f'a {16} on the ACT maps to a {act_to_sat[16]} on the SAT')
print(f'a {35} on the ACT maps to a {act_to_sat[35]} on the SAT')

a 28 on the ACT maps to a 657.9999999999999 on the SAT
a 16 on the ACT maps to a 445.0 on the SAT
a 35 on the ACT maps to a 782.25 on the SAT


We are going to read in a SAT to ACT conversion chart from [The Princeton Review]("https://www.princetonreview.com/college-advice/act-to-sat-conversion")

In [124]:
ACT_SAT = pd.read_html("https://www.princetonreview.com/college-advice/act-to-sat-conversion")

In [125]:
conversion_df = ACT_SAT[0]

In [126]:
conversion_df = conversion_df.drop(["Unnamed: 2", "Unnamed: 5"], axis=1)

In [130]:
conversion_df.head(36)

SAT  ACT  SAT.1  ACT.1  SAT.2  ACT.2
0   1600   36   1250     26    900     17
1   1590   35   1240     26    890     16
2   1580   35   1230     25    880     16
3   1570   35   1220     25    870     16
4   1560   35   1210     25    860     16
5   1550   34   1200     25    850     15
6   1540   34   1190     24    840     15
7   1530   34   1180     24    830     15
8   1520   34   1170     24    820     15
9   1510   33   1160     24    810     15
10  1500   33   1150     23    800     14
11  1490   33   1140     23    790     14
12  1480   32   1130     23    780     14
13  1470   32   1120     22    770     14
14  1460   32   1110     22    760     14
15  1450   32   1100     22    750     13
16  1440   31   1090     21    740     13
17  1430   31   1080     21    730     13
18  1420   31   1070     21    720     13
19  1410   30   1060     21    710     12
20  1400   30   1050     20    700     12
21  1390   30   1040     20    690     12
22  1380   29   1030     20    680     12
23  1370   29   1020     20    670     12
24  1360   29   1010     19    660     12
25  1350   29   1000     19    650     12
26  1340   28    990     19    640     12
27  1330   28    980     19    630     12
28  1320   28    970     18    620     11
29  1310   28    960     18    610     11
30  1300   27    950     18    600     11
31  1290   27    940     18    590     11
32  1280   27    930     17    580     11
33  1270   26    920     17    570     11
34  1260   26    910     17    560     11

In [93]:
def convert_to_act(x):
    if x < 360:
        return 12
    elif x < 380:
        return 13
    elif x < 405:
        return 14
    elif x < 430:
        return 15
    elif x < 450:
        return 16
    elif x < 470:
        return 17
    elif x < 480:
        return 18
    elif x < 510:
        return 19
    elif x < 1060:
        return 20
    elif x < 1100:
        return 21
    elif x < 1130:
        return 22
    elif x < 1160:
        return 23
    elif x < 1200:
        return: 24
    elif x < 1240:
        return 25
    elif x < 1280:
        return 26
    elif x < 

In [94]:
conversion1 = conversion_df[['ACT', 'SAT']]
conversion2 = conversion_df[['ACT.1', 'SAT.1']]
conversion3 = conversion_df[['ACT.2', 'SAT.2']]

In [95]:
conversion2.columns = ['ACT', 'SAT']
conversion3.columns = ['ACT', 'SAT']

In [96]:
conversion_df = pd.concat([conversion1, conversion2, conversion3])

In [97]:
conversion_df.set_index('ACT', inplace=True)

In [98]:
groupby = conversion_df.reset_index().groupby('ACT')

In [99]:
conversion_df = groupby.agg('mean')

In [100]:
conversion_df['SAT'] = conversion_df['SAT'] / 2

In [101]:
act_sat = conversion_df['SAT'].to_dict()

In [102]:
sat_act = {v: i for i,v in act_sat.items()}

In [103]:
sat_act

{295.0: 11,
 335.0: 12,
 367.5: 13,
 390.0: 14,
 415.0: 15,
 437.5: 16,
 457.5: 17,
 477.5: 18,
 497.5: 19,
 517.5: 20,
 537.5: 21,
 555.0: 22,
 570.0: 23,
 587.5: 24,
 607.5: 25,
 627.5: 26,
 645.0: 27,
 662.5: 28,
 682.5: 29,
 700.0: 30,
 715.0: 31,
 732.5: 32,
 750.0: 33,
 767.5: 34,
 787.5: 35,
 800.0: 36}

In [104]:
act_sat

{11: 295.0,
 12: 335.0,
 13: 367.5,
 14: 390.0,
 15: 415.0,
 16: 437.5,
 17: 457.5,
 18: 477.5,
 19: 497.5,
 20: 517.5,
 21: 537.5,
 22: 555.0,
 23: 570.0,
 24: 587.5,
 25: 607.5,
 26: 627.5,
 27: 645.0,
 28: 662.5,
 29: 682.5,
 30: 700.0,
 31: 715.0,
 32: 732.5,
 33: 750.0,
 34: 767.5,
 35: 787.5,
 36: 800.0}

Now I have my conversions saved as act_sat and sat_act. If A school has either one of these values filled in we can map the rest of the scores with the following conversion dictionaries

In [105]:
(final_df.filter(regex="(SAT|ACT).*isNaN").sum(axis=1) == 7).sum()

1051

At the end we should have 1051 null values in all test score columns

In [106]:
# ACT composite has the least amount of null values with 1096
final_df.filter(regex="ACT Comp").iloc[:, 0].isna().sum()

1096

In [107]:
# first define ACT column
act_col = final_df.filter(regex="ACT Comp").columns[0]

In [108]:
# make a column of ACT to SAT score conversions
final_df["converted_act"] = final_df[act_col].fillna(0).astype(int).map(act_sat)

In [109]:
# fill SAT scores that are missing with conversions from ACT Scores
sat_math = final_df.filter(regex="SAT Math").columns[0]

In [110]:
# rename the sat_math column
final_df.rename({sat_math: "sat_math75"}, axis=1, inplace=True)

In [111]:
final_df["sat_math75"].describe() #min is 400, max is 800 after we map we will replace 0's back with NA values

count    1175.000000
mean      587.587234
std        74.016774
min       400.000000
25%       540.000000
50%       580.000000
75%       630.000000
max       800.000000
Name: sat_math75, dtype: float64

In [112]:
# find indices where math is null and but ACT Composite is not
filter1 = final_df["ACT Composite 75th percentile score_isNaN"] == 0
filter2 = final_df["SAT Math 75th percentile score_isNaN"] == 1

# save these indices as ilocs
ilocs = final_df[filter1 & filter2].index

# map the indices where SAT Math is null to the converted_act column, which are converted into sat values
final_df.loc[ilocs, "sat_math75"] = final_df.loc[ilocs, "converted_act"]

# lets view the null values in math sat scores
final_df["sat_math75"].isna().sum()

1051

In [113]:
# Lets repeat this process for reading 
filter1 = final_df["ACT Composite 75th percentile score_isNaN"] == 0
filter2 = final_df["SAT Critical Reading 75th percentile score_isNaN"] == 1

# save these indices as ilocs
ilocs = final_df[filter1 & filter2].index

# map the indices where SAT Math is null to the converted_act column, which are converted into sat values
final_df.loc[ilocs, "SAT Critical Reading 75th percentile score"] = final_df.loc[ilocs, "converted_act"]

# lets view the null values in math sat scores
final_df["SAT Critical Reading 75th percentile score"].isna().sum()

1051

In [114]:
# Lets repeat this process one more time for writing

filter1 = final_df["ACT Composite 75th percentile score_isNaN"] == 0
filter2 = final_df["SAT Writing 75th percentile score_isNaN"] == 1

# save these indices as ilocs
ilocs = final_df[filter1 & filter2].index

# map the indices where SAT Math is null to the converted_act column, which are converted into sat values
final_df.loc[ilocs, "SAT Writing 75th percentile score"] = final_df.loc[ilocs, "converted_act"]

# lets view the null values in math sat scores
final_df["SAT Writing 75th percentile score"].isna().sum()

1078

In [115]:
# Lets repeat this process one more time for writing

filter1 = final_df["SAT Critical Reading 75th percentile score_isNaN"] == 0
filter2 = final_df["SAT Writing 75th percentile score_isNaN"] == 1

# save these indices as ilocs
ilocs = final_df[filter1 & filter2].index

# map the indices where SAT Math is null to the converted_act column, which are converted into sat values
final_df.loc[ilocs, "SAT Writing 75th percentile score"] = final_df.loc[ilocs, "SAT Critical Reading 75th percentile score_isNaN"]

# lets view the null values in math sat scores
final_df["SAT Writing 75th percentile score"].isna().sum()

1051

In [116]:
final_df["converted_sat"] = final_df["SAT Critical Reading 75th percentile score"].fillna(0).astype(int).map(sat_act)

In [117]:
final_df["SAT Critical Reading 75th percentile score"].describe()
print(sat_act)

{295.0: 11, 335.0: 12, 367.5: 13, 390.0: 14, 415.0: 15, 437.5: 16, 457.5: 17, 477.5: 18, 497.5: 19, 517.5: 20, 537.5: 21, 555.0: 22, 570.0: 23, 587.5: 24, 607.5: 25, 627.5: 26, 645.0: 27, 662.5: 28, 682.5: 29, 700.0: 30, 715.0: 31, 732.5: 32, 750.0: 33, 767.5: 34, 787.5: 35, 800.0: 36}


In [118]:
# Lets repeat this process except fill for ACT scores

filter1 = ~final_df["sat_math75"].isna()
filter2 = final_df["ACT Composite 75th percentile score"].isna()

ilocs = final_df[filter1 & filter2].index


final_df.loc[ilocs, "ACT Composite 75th percentile score"] = final_df.loc[ilocs, "converted_sat"]

final_df["ACT Composite 75th percentile score"].isna().sum()

1092

In [119]:
final_df.loc[ilocs, "converted_sat"]

UnitID
139199     NaN
139630     NaN
141361     NaN
161022     NaN
161518     NaN
161873     NaN
162283     NaN
163286     NaN
163295     NaN
164155     NaN
164614     NaN
166638    23.0
167455     NaN
183062     NaN
183071     NaN
184773     NaN
184782     NaN
185129     NaN
185590     NaN
186371     NaN
186380     NaN
186399     NaN
186618     NaN
187444     NaN
190512     NaN
190549    23.0
190558     NaN
190567     NaN
190594     NaN
190600     NaN
190637     NaN
190664     NaN
190691     NaN
192925     NaN
194116     NaN
195173     NaN
199698     NaN
209825     NaN
213826     NaN
216931    23.0
233921     NaN
243106     NaN
243142    23.0
243151     NaN
243221     NaN
Name: converted_sat, dtype: float64

In [120]:
act_na = final_df.filter(regex="ACT Comp").columns[1]
sat_na = final_df.filter(regex="SAT").columns[-1]

In [121]:
len(np.where(final_df.filter(regex="SAT").iloc[:, 0].isna()))

1

In [122]:
filt1 = final_df[act_na] == 0
